cura/Machines/MachineNode.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) 2019 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from typing import Dict, List
+
+from UM.Logger import Logger
+from UM.Signal import Signal
+from UM.Util import parseBool
+from UM.Settings.ContainerRegistry import ContainerRegistry  # To find all the variants for this machine.
+from cura.Machines.ContainerNode import ContainerNode
+from cura.Machines.QualityChangesGroup import QualityChangesGroup  # To construct groups of quality changes profiles that belong together.
+from cura.Machines.QualityGroup import QualityGroup  # To construct groups of quality profiles that belong together.
+from cura.Machines.QualityNode import QualityNode
+from cura.Machines.VariantNode import VariantNode
+
+##  This class represents a machine in the container tree.
+#
+#   The subnodes of these nodes are variants.
+class MachineNode(ContainerNode):
+    def __init__(self, container_id: str) -> None:
+        super().__init__(container_id)
+        self.variants = {}  # type: Dict[str, VariantNode] # Mapping variant names to their nodes.
+        self.global_qualities = {}  # type: Dict[str, QualityNode] # Mapping quality types to the global quality for those types.
+        self.materialsChanged = Signal()  # Emitted when one of the materials underneath this machine has been changed.
+
+        container_registry = ContainerRegistry.getInstance()
+        try:
+            my_metadata = container_registry.findContainersMetadata(id = container_id)[0]
+        except IndexError:
+            Logger.log("Unable to find metadata for container %s", container_id)
+            my_metadata = {}
+        # Some of the metadata is cached upon construction here.
+        # ONLY DO THAT FOR METADATA THAT DOESN'T CHANGE DURING RUNTIME!
+        # Otherwise you need to keep it up-to-date during runtime.
+        self.has_materials = parseBool(my_metadata.get("has_materials", "true"))
+        self.has_variants = parseBool(my_metadata.get("has_variants", "false"))
```

Comment:
Note that some machines can have options, which can change fields such as `has_variants` in the `GlobalStack`. For example, with UM2, you can check `Olsson Block`, which will set `has_variants = True` in the `GlobalStack`'s metadata. See:

https://github.com/Ultimaker/Cura/blob/60c7a3d3b72cc41a5c85dd159705c1f80b5c86e6/plugins/UltimakerMachineActions/UM2UpgradeSelection.py#L46-L71

Reply:
Fixed this in CURA-6775.